pkg/operator/ceph/object/bucket/provisioner.go

```diff
@@ -568,75 +572,47 @@ func (p *Provisioner) populateDomainAndPort(sc *storagev1.StorageClass) error {
 
 // Check for additional options mentioned in OBC and set them accordingly
 func (p *Provisioner) setAdditionalSettings(options *apibkt.BucketOptions) error {
-	var maxObjectsInt64 int64 = -1
-	var maxSizeInt64 int64 = -1
-	var err error
-	var quotaEnabled bool
-
-	maxObjects := MaxObjectQuota(options.ObjectBucketClaim.Spec.AdditionalConfig)
-	if maxObjects != "" {
-		quotaEnabled = true
-
-		maxObjectsInt64, err = toInt64(maxObjects)
-		if err != nil {
-			return errors.Wrapf(err, "failed to parse maxObjects quota for user %q", p.cephUserName)
-		}
-	}
-
-	maxSize := MaxSizeQuota(options.ObjectBucketClaim.Spec.AdditionalConfig)
-	if maxSize != "" {
-		quotaEnabled = true
-
-		maxSizeInt64, err = toInt64(maxSize)
-		if err != nil {
-			return errors.Wrapf(err, "failed to parse maxSize quota for user %q", p.cephUserName)
-		}
+	additionalConfig, err := additionalConfigSpecFromMap(options.ObjectBucketClaim.Spec.AdditionalConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to process additionalConfig")
 	}
 
 	objectUser, err := p.adminOpsClient.GetUser(p.clusterInfo.Context, admin.User{ID: p.cephUserName})
 	if err != nil {
 		return errors.Wrapf(err, "failed to fetch user %q", p.cephUserName)
 	}
+	currentQuota := objectUser.UserQuota
 
 	// enable or disable quota for user
-	if *objectUser.UserQuota.Enabled != quotaEnabled {
-		err = p.adminOpsClient.SetUserQuota(p.clusterInfo.Context, admin.QuotaSpec{UID: p.cephUserName, Enabled: &quotaEnabled})
-		if err != nil {
-			return errors.Wrapf(err, "failed to set user %q quota enabled=%v for obc", p.cephUserName, quotaEnabled)
-		}
-	}
+	quotaEnabled := (additionalConfig.maxObjects != nil) || (additionalConfig.maxSize != nil)
 
-	if !quotaEnabled {
-		// no need to process anything else if quotas are disabled
-		return nil
-	}
+	var needSync bool
 
-	if *objectUser.UserQuota.MaxObjects != maxObjectsInt64 {
-		err = p.adminOpsClient.SetUserQuota(p.clusterInfo.Context, admin.QuotaSpec{UID: p.cephUserName, MaxObjects: &maxObjectsInt64})
-		if err != nil {
-			return errors.Wrapf(err, "failed to set MaxObjects=%v to user %q", maxObjectsInt64, p.cephUserName)
-		}
+	if *currentQuota.Enabled != quotaEnabled {
+		needSync = true
+	}
+	if (currentQuota.MaxObjects != nil) && (additionalConfig.maxObjects != nil) && (*currentQuota.MaxObjects != *additionalConfig.maxObjects) {
+		needSync = true
+	}
+	if (currentQuota.MaxSize != nil) && (additionalConfig.maxSize != nil) && (*currentQuota.MaxSize != *additionalConfig.maxSize) {
+		needSync = true
 	}
```

Comment:
Reconciliation logic tends to work best (and have fewer bugs throughout long-term maintenance) when each iteration builds up the new config in as much of a 1:1 way as possible. Logic can then compare whether the existing state of the world diverges and needs updating using simple logic.

Using bools to keep track of whether syncing needs to happen is sometimes necessary, but it's best avoided when possible because it's too easy for those-who-come-after to miss important logic.

Instead of rendering the user desired state into an intermediate type (`additionalConfigSpec`), can we render user desired state into the same type as `currentQuota`, and then compare them with `DeepEqual` (or some other comparison logic)?

Reply:
I would like to keep `additionalConfigSpec` and possibly add it into `Provisioner` down the road .  That allows the interpretation of the `additonalConfig` map in one function instead of having having conversion statements salted around.  I think we could end up with ~7 `additionlConfig` keys.

I considered trying to compare [QuotaSpec](https://pkg.go.dev/github.com/ceph/go-ceph/rgw/admin#QuotaSpec)s and I didn't do it since there are a few fields not being managed.  Maybe I can make it work by using a deep copy?

Reply:
I'd prefer not write code that is going to be hard to maintain based on what might or might not happen in the future. Code can always be improved, but code maintenance is eternal. 

I have observed that the `if X; needUpdate=true; fi; if Y; needUpdate=true; fi; if Z; needUpdate=true; fi` code tends to develop bugs in the long term when different people get involved. I really want to avoid that if possible.

Reply:
This should also have been short circuited on `needSync` or written as an if/else if from an efficiency perspective.

Reply:
I don't really care to nitpick over nanoseconds. But I care a lot about how resilient the code is to people making changes in the future and potentially missing small thing like this. As a maintainer, this is something that I see regularly, and I try to avoid patterns that don't insulate against change.

Reply:
I've reworked the comparison to use QuotaSpec in a way that I think won't be brittle to changes in unmanaged fields.